src/group_theory/index.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2021 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+-/
+
+import group_theory.coset
+import set_theory.cardinal
+
+/-!
+# Index of a Subgroup
+
+In this file we define the index of a subgroup, and prove several divisibility properties.
+
+## Main definitions
+
+- `H.index` : the index of `H : subgroup G`
+
+# Main results
+
+- `index_mul_card` : `H.index * fintype.card H = fintype.card G`
+- `index_dvd_card` : `H.index ∣ fintype.card G`
+- `index_eq_mul_of_le` : If `H ≤ K`, then `H.index = K.index * (H.subgroup_of K).index`
+- `index_dvd_of_le` : If `H ≤ K`, then `K.index ∣ H.index`
+
+-/
+
+namespace subgroup
+
+variables {G : Type*} [group G] (H : subgroup G)
+
+/-- The index of a subgroup as a natural number -/
+noncomputable def index : ℕ :=
+(cardinal.mk (quotient_group.quotient H)).to_nat
+
+lemma index_eq_card [fintype (quotient_group.quotient H)] :
+  H.index = fintype.card (quotient_group.quotient H) :=
+cardinal.mk_to_nat_eq_card
+
+lemma index_mul_card [fintype G] [fintype H] :
+  H.index * fintype.card H = fintype.card G :=
+begin
+  classical,
+  rw H.index_eq_card,
+  convert H.card_eq_card_quotient_mul_card_subgroup.symm,
+end
+
+lemma index_dvd_card [fintype G] : H.index ∣ fintype.card G :=
+begin
+  classical,
+  rw H.index_eq_card,
+  convert H.card_quotient_dvd_card,
+end
+
+variables {H} {K : subgroup G}
+
+/-- If `H ≤ K` then `G/H ≃ G/K × K/H` -/
+def quotient_equiv_prod_of_le' (h_le : H ≤ K) (f : quotient_group.quotient K → G)
+  (hf : function.right_inverse f quotient_group.mk) :
+  quotient_group.quotient H ≃
+  quotient_group.quotient K × quotient_group.quotient (H.subgroup_of K) :=
+{ to_fun := λ a, ⟨a.map' id (by exact λ b c h, h_le h),
```

Comment:
Does `@[simps]` generate anything useful? If not, can you manually add a lemma saying that `quotient_equiv_prod_of_le' ... (mk' a) = (mk' a, _)`